### DISCUSSION TITLE
Shortcodes or similar extension mechanism

### DISCUSSION THREAD

**Discussion by simonbates** [mkdocs/mkdocs#3182] [Q&A]
I have recently been looking at documentation options for a new project and I've been trying MkDocs out.

I like it very much. In particular, I really like that it's possible to generate HTML that can be read offline from the filesystem as well as served from a web server. This is something that I'd like to be able to do with the documentation I'm writing: host it on the web, but also have it be possible that it can be read locally without a web server. There are lots of other awesome things about MkDocs, such as internal link checking, that makes it a good fit for what I'm looking for, but the ability to generate docs that can be read offline is a big one. And it seems fairly rare in static site generators.

I'd like to be able to extend/customize the markup that I'm writing. A concrete example is I have code to generate SVG diagrams that I'd like to be able to include inline in the documentation. And I will have recurring types of content that I'd like to be able to render and style consistently.

I've been looking at the extension options that I was able to find: adding Python-Markdown extensions and MkDocs plugins. I was thinking that it would be nice to have a simpler extension mechanism for site-specific custom markup. Similar to [shortcodes in 11ty](https://www.11ty.dev/docs/shortcodes/) or [shortcodes in Hugo](https://gohugo.io/templates/shortcode-templates/) or [React components in Docusaurus](https://docusaurus.io/docs/markdown-features/react).

I was wondering if such an extension mechanism was something that would be of interest/value to MkDocs? I'm very new to the project and I don't know if this is something that has been discussed in the past, or if it would be considered in scope for MkDocs. I would be happy to help if it's something that would be useful as it's something that I'd like to have for my own usage.

I can think of a number of ways of implementing such an extension mechanism, such as:

- A Markdown preprocessor using a custom syntax that we design (which could be implemented as a plugin)
- A Markdown preprocessor which runs the page content through Jinja as a template before Markdown (I believe that this is the approach that 11ty takes)
- A Python-Markdown extension

I'm too new to the project to know which might fit best.

This morning I wrote up a quick sketch to illustrate the sort of thing that I'm thinking of:

https://github.com/simonbates/mkdocs-shortcodes-sketch

This sketch provides an API similar to 11ty: shortcode handlers are implemented in code (Python for MkDocs, rather than JavaScript for 11ty). A simple example to wrap the content of a `note` shortcode in `<div class="note">...</div>` looks like:

```python
add_shortcode('note', lambda content : '<div class="note">\n' + content + '\n</div>')
```

And it's used like:

```text
=begin note
note
content
=end
```

See:

- https://github.com/simonbates/mkdocs-shortcodes-sketch/blob/main/shortcodes.py
- https://github.com/simonbates/mkdocs-shortcodes-sketch/blob/main/docs/index.md

My sketch is implemented as a Markdown preprocessor MkDocs plugin (the first option above). This is just a super quick example of the sort of thing that I'm thinking of. Definitely not a final proposal.

**Comment by simonbates** [comment]
I just found the GitHub Discussions for MkDocs and in the topic [Status of the project](https://github.com/mkdocs/mkdocs/discussions/3145) it sounds like MkDocs is considered feature complete. So I'm imagining a shortcodes, or a similar, extension mechanism wouldn't be something that the team would want to add to the main project at this time. Is that a reasonable reading of the situation? Probably then for me, the best path forward would be to make something for my own needs as a plugin/hooks. And then if it's something that would be more generally useful, I could publish it as a plugin so that others could use it.

**Comment by pawamoy** [comment]
Hey @simonbates!

Thanks for your well written questions and use-case, that was pleasant to read :slightly_smiling_face: 

I think you'll be interested in the [Blocks extension](https://facelessuser.github.io/pymdown-extensions/extensions/blocks/)  of [PyMdown Extensions](https://facelessuser.github.io/pymdown-extensions/). This extension allows to define custom blocks and to use them like this:

```
/// note | Did you know?
You can create a note with Blocks!
///

/// my_custom_block | some attribute
Contents
///
```

You might also find [Markdown Exec](https://pawamoy.github.io/markdown-exec/) interesting, since it would allow you to inline your code that generates SVG directly in your Markdown page:

````
```python exec="1"
print(your_function_that_generates_svg_diagrams())
```
````

It only supports Python and bash/zsh/sh right now, but it's possible to add support for more languages/interpreters.
